sw/otbn/crypto/p256.s

```diff
@@ -2254,6 +2325,256 @@ boolean_to_arithmetic:
+ * @param[out] w20: boolean masked value x', such that x = x' ^ r
+ *
+ * clobbered registers: w1 - w6, w11, w12, w18, w20 - w27, and w29
+ * clobbered flag groups: FG0
+ */
+arithmetic_to_boolean_mod:
+  /* First step: calculate A2B from reduced values. */
+
+  /* Save inputs for second A2B execution.
+     w24 <= w19 = r
+     w25 <= w11 = A */
+  bn.mov    w24, w19
+  bn.mov    w25, w11
+
+  /* Expand inputs r and A (w19 and w11) to 257-bit values [w19,w18]
+     and [w12,w11] and prepare input for 257-bit A2B function.
+     w18 <= w19
+     w19 <= w31
+     w11 <= w11 -> obsolete
+     w12 <= w31 */
+  bn.mov    w18, w19
+  bn.mov    w19, w31
+  bn.mov    w12, w31
+
+  /* Call 257-bit A2B function.
+     [w21,w20] <= x' */
+  jal       x1, arithmetic_to_boolean
+
+  /* Save intermediate result of reduced inputs.
+     w26 <= w20 = x' (lower part)
+     w27 <= w21 = x' (upper part) */
+  bn.mov    w26, w20
+  bn.mov    w27, w21
+
+  /* Second step: calculate A2B from unreduced values. */
+
+  /* Restore and expand inputs r and A (w19 and w11) to 257-bit
+     values [w19,w18] and [w12,w11] and prepare input for
+     257-bit A2B function.
+     w18 <= w24
+     w19 <= w31
+     w11 <= w25
+     w12 <= w31 */
+  bn.mov    w18, w24
+  bn.mov    w19, w31
+  bn.mov    w11, w25
+  bn.mov    w12, w31
+
+  /* Get field modulus p.
+     w29 <= MOD() */
+  bn.wsrr   w29, 0x00 /* MOD */
+
+  /* Convert input A ([w12,w11]) to an unreduced value
+     in the 2^257 domain. For this add (2^257 - p) to A.
+     [w12,w11] <= [w12,w11] + 2^257 - w29 = A + 2^257 - p
+     w12 <= w12 + 0x2 = A + 2^257
+            -> equal to addition of 2^257
+               (w11 doesn't need to be touched)
+     [w12,w11] <= [w12,w11] - w29 = (A + 2^257) - p */
+  bn.addi   w12, w12, 0x2
+  bn.sub    w11, w11, w29
+  bn.subb   w12, w12, w31
+
+  /* Call 257-bit A2B function.
+     [w21,w20] <= x' */
+  jal       x1, arithmetic_to_boolean
+
+  /* Restore initial mask input of w19 for consistency
+     in calling functions.
+     w19 <= w24 */
+  bn.mov    w19, w24
+
+  /* Check MSB (carry bit) of second A2B result for true or false. */
+  bn.cmp    w21, w31 /* w21 can only be 0x1 or 0x0 */
+
+  /* Read the FG0.Z flag (bit 3).
+     x2 <= FG0.Z */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 8
+  srli      x2, x2, 3
+
+  /* If FG0.Z flag is 0 (-> bn.cmp was not 0 -> MSB true),
+     return reduced A2B result. */
+  beq       x2, x0, arithmetic_to_boolean_mod_ret_reduced
+
+  /* Return result of unreduced A2B computation (second result). */
+  arithmetic_to_boolean_mod_ret_unreduced:
+  /* Result is still in w20, but because of constant time
+     requirement this is done anyway.
+     w20 <= w20 */
+  bn.mov    w20, w20
+
+  ret
+
+  /* Return result of reduced A2B computation (first result). */
+  arithmetic_to_boolean_mod_ret_reduced:
+  /* Restore previous result to output WDR.
+     w20 <= w26 */
+  bn.mov    w20, w26
```

Comment:
I think you can do this whole operation  with `bn.sel`:
```suggestion
bn.sel    w20, w20, w26, FG0.Z
```

Reply:
That's actually a very good improvement. I changed it and it still works as intended. Thank you!